src/core/pointcloud/qgsremotecopcpointcloudindex.cpp

```diff
@@ -0,0 +1,373 @@
+                         --------------------
+    begin                : March 2021
+    copyright            : (C) 2021 by Belgacem Nedjima
+    email                : belgacem dot nedjima at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsremotecopcpointcloudindex.h"
+
+#include <QFile>
+#include <QFileInfo>
+#include <QDir>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QTime>
+#include <QtDebug>
+#include <QQueue>
+#include <QTimer>
+
+#include "qgseptdecoder.h"
+#include "qgscoordinatereferencesystem.h"
+#include "qgspointcloudrequest.h"
+#include "qgspointcloudattribute.h"
+#include "qgslogger.h"
+#include "qgsfeedback.h"
+#include "qgsmessagelog.h"
+
+#include "qgstiledownloadmanager.h"
+#include "qgsblockingnetworkrequest.h"
+#include "qgslazdecoder.h"
+#include "qgsfileutils.h"
+#include "qgsapplication.h"
+#include "qgscopcpointcloudblockrequest.h"
+#include "qgspointcloudexpression.h"
+#include "qgsnetworkaccessmanager.h"
+
+///@cond PRIVATE
+
+QgsRemoteCopcPointCloudIndex::QgsRemoteCopcPointCloudIndex() : QgsCopcPointCloudIndex()
+{
+
+}
+
+QgsRemoteCopcPointCloudIndex::~QgsRemoteCopcPointCloudIndex() = default;
+
+QList<IndexedPointCloudNode> QgsRemoteCopcPointCloudIndex::nodeChildren( const IndexedPointCloudNode &n ) const
+{
+  fetchNodeHierarchy( n );
+
+  mHierarchyMutex.lock();
+  Q_ASSERT( mHierarchy.contains( n ) );
+  QList<IndexedPointCloudNode> lst;
+  const int d = n.d() + 1;
+  const int x = n.x() * 2;
+  const int y = n.y() * 2;
+  const int z = n.z() * 2;
+  mHierarchyMutex.unlock();
+
+  for ( int i = 0; i < 8; ++i )
+  {
+    int dx = i & 1, dy = !!( i & 2 ), dz = !!( i & 4 );
+    const IndexedPointCloudNode n2( d, x + dx, y + dy, z + dz );
+    if ( fetchNodeHierarchy( n2 ) && mHierarchy[n] > 0 )
+      lst.append( n2 );
+  }
+  return lst;
+}
+
+void QgsRemoteCopcPointCloudIndex::load( const QString &url )
+{
+  mUrl = QUrl( url );
+
+  mIsValid = loadHeader();
+  if ( !mIsValid )
+    return;
+
+  fetchHierarchyPage( mCopcInfoVlr.root_hier_offset, mCopcInfoVlr.root_hier_size );
+}
+
+
+bool QgsRemoteCopcPointCloudIndex::loadHeader()
+{
+  // Request header data
+  {
+    QNetworkRequest nr( mUrl );
+    nr.setAttribute( QNetworkRequest::HttpPipeliningAllowedAttribute, true );
```

Comment:
is this required? does it have some effect here? as we're using blocking network request here, it does not seem like we would be able to put multiple requests into the pipeline?

Reply:
Not really I was experimenting with that flag a bit and forgot it there